src/rt/lifetime.d

```diff
@@ -281,15 +283,7 @@ bool __setArrayAllocLength(ref BlkInfo info, size_t newlength, bool isshared, co
         {
             if(isshared)
             {
-                try synchronized(typeid(ArrayAllocLengthLock))
-                {
-                    if(*length == cast(ubyte)oldlength)
-                        *length = cast(ubyte)newlength;
-                    else
-                        return false;
-                }
-                catch (Throwable t)
-                    assert(0, "Failed to synchronize.");
+                return cas(cast(shared)length, cast(ubyte)oldlength, cast(ubyte)newlength);
```

Comment:
Well, I suppose that makes sense, but I don't understand the point of it. The entire point of weak pure functions is to be called from strong-pure functions. Clearly, a strong-pure function cannot pass in shared data.

I don't agree that locking a mutex is any different than touching shared data. But accessing the global typeid should not work (is there a bug report for that?)